tools/isobuild/linker.js

```diff
@@ -145,32 +150,121 @@ _.extend(Module.prototype, {
     // Prologue
     var chunks = [];
 
+    const result = {
+      // This object will be populated with .source, .servePath,
+      // .sourceMap, and (optionally) .exportsName properties before being
+      // returned from this method in a singleton array.
+      servePath: self.combinedServePath,
+    };
+
     // Emit each file
-    _.each(self.files, function (file) {
-      if (!_.isEmpty(chunks)) {
-        chunks.push("\n\n\n\n\n\n");
+    if (self.useMeteorInstall) {
+      const tree = {};
+
+      _.each(self.files, function (file) {
+        if (file.bare) {
+          // Bare files will be added in between the synchronous require
+          // calls below.
+          return;
+        }
+
+        const parts = file.installPath.split("/");
+        let t = tree;
+        _.each(parts, function (part, i) {
+          const isLastPart = i === parts.length - 1;
+          t = _.has(t, part)
+            ? t[part]
+            : t[part] = isLastPart ? file : {};
+        });
+      });
+
+      let moduleCount = 0;
+
+      function walk(t) {
+        if (t instanceof File) {
+          ++moduleCount;
+          chunks.push(t.getPrelinkedOutput({
+            sourceWidth,
+            noLineNumbers: self.noLineNumbers
+          }));
+        } else if (_.isObject(t)) {
+          chunks.push("{");
+          const keys = _.keys(t);
+          _.each(keys, (key, i) => {
+            chunks.push(JSON.stringify(key), ":");
+            walk(t[key]);
+            if (i < keys.length - 1) {
+              chunks.push(",");
+            }
+          });
+          chunks.push("}");
+        }
       }
-      chunks.push(file.getPrelinkedOutput({
-        sourceWidth: sourceWidth,
-        noLineNumbers: self.noLineNumbers
-      }));
-    });
+
+      const chunksLengthBeforeWalk = chunks.length;
+
+      // The tree of nested directories and module functions built above
+      // allows us to call meteorInstall just once to install everything.
+      chunks.push("var require = meteorInstall(");
+      walk(tree);
+      chunks.push(");");
+
+      if (moduleCount === 0) {
+        // If no files were actually added to the chunks array, roll back
+        // to before the `var require = meteorInstall(` chunk.
+        chunks.length = chunksLengthBeforeWalk;
+      }
+
+      // Now that we have installed everything in this package or
+      // application, immediately require the non-lazy modules and
+      // evaluate the bare files.
```

Comment:
The output of `file.getPrelinkedOutput` is different now. In the original code, we could just output the results of `file.getPrelinkedOutput` for both `bare` and non-`bare` files, but now the output for non-`bare` files is a blob of `meteorInstall` code, so it's not possible to insert `bare` code in between the bundled modules. More importantly, `meteorInstall` doesn't actually evaluate any module code; that happens when we call `require`. If it's important for a `bare` file to be evaluated between two modules, then we need to output code that `require`s the first module, then code for the `bare` module, then code that `require`s the second module.

I would like to see `bare` files (and the `self.useGlobalNamespace` thing above) go away entirely, but that would probably break back compat.